### ISSUE TITLE
Allow setting of inter_op_num_threads, intra_op_num_threads and intra_op.allow_spinning

### ISSUE THREAD

**Issue by RJKeevil** [yalue/onnxruntime_go#10]
Hi, i'm looking at performance tuning of Onnx Runtime for my model, reading online (e.g. https://www.inworld.ai/blog/reducing-cpu-usage-in-machine-learning-model-inference-with-onnx-runtime) it seems like inter_op_num_threads, intra_op_num_threads  and intra_op.allow_spinning and the first things to try.  Looks like calling run on a session concurrently is working nicely, so id like to try parallelising at the call level, and restricting each call to a single thread each.  At the moment I don't seem to be able to set these options, either via argument or env var. Could this be added?

**Comment by yalue** [comment]
Hi, thank you for the interest in the library, and yes, I apologize but the library currently does not expose these settings to Go code.  I am actively working on adding features that will allow for doing so, but I do not yet know how long it will take me to finish!  I hope to be done in the next two or three weeks.

However, in the meantime, I somewhat addressed this topic in Issue #2.  If you're able to modify the C code in this repo, you can add the modifications you need to the `CreateSession` function in `onnxruntime_wrapper.c` pretty easily.  You can see the snippet I posted in issue #2 where I added a few lines to set the `inter_op_num_threads` and `intra_op_num_threads` options prior to creating the session in C.  I hope that can work for you for now; I'd imagine you can combine the snippet I posted there with some calls to `getenv` to set the options based on environment variables, for example.

**Comment by yalue** [comment]
Hello again @RJKeevil, thought I'd mention here that I have now added a "proper" API that exposes the C `SetInterOpNumThreads` and `SetIntraOpNumThreads` APIs within Go.  It's part of the changes added with commit 7bf5a6e (PR #14).  If you still need a solution, you should be able to do this within Go code now like this (in a snippet I copied from another issue):

```go
// Note: Omitting error checking

// Create a SessionOptions object and configure the # of threads.
options, _ := ort.NewSessionOptions()
defer options.Destroy()
options.SetIntraOpNumThreads(1)
options.SetInterOpNumThreads(1)

// Provide the options pointer when creating a new session
session, _ := ort.NewAdvancedSession("network.onnx",
    []string{"InputTensor"}, []string{"OutputTensor"},
    []ArbitraryTensor{inputTensor}, []ArbitraryTensor{outputTensor},
    options)
defer session.Destroy()

// ...
```

**Comment by RJKeevil** [comment]
Hi @yalue , thanks for the nudge, I tried to upgrade to this but hit a snag.  Whilst the mismatched input/output types are no longer an issue with the new AdvancedSession, the ability to run Onnx on separate input/output tensor instances is still critical to be able to thread properly, as different threads need to work independently on different input/output data. I tried going back to the DynamicSession in this version, but it doesnt include the new options argument, so i still need my fork to add the thread arguments.

I think the cleanest answer is to keep supporting the DynamicSession (i.e. dont deprecate it), remove the type declarations to bring it in line with the AdvancedSession API, and add the new options argument to it.  I know this isnt backwards compatible but I think it would be messy to add another session type, and as it was added recently there isnt too much legacy to worry about.  Let me know what you think!

**Comment by yalue** [comment]
Ah, yeah, I bet I know what the issue is: it was actually avoided by the way in which you allocated new `*OrtValue` arrays in your implementation.  Doing things your way (rather than caching a pre-allocated list of `OrtValue`s) will have meant that concurrent calls to `Run` wouldn't have been clobbering one another's values.  Stupid mistake on my part, and thank you for bearing with me while I address this problem.  Should be easy enough to fix, which I will do so shortly.

I probably should also add a race condition unit test, for using with `go test -race`, and add to the comments that concurrent calls to dynamic session's Run() functions are allowed.

**Comment by yalue** [comment]
This should hopefully be fixed again with commit e73094ac6b1.  Please let me know if it works again, and sorry for breaking your code!

**Comment by RJKeevil** [comment]
Yes this works perfectly now! Thanks for the fast turnaround.  No need to apologise, i'm very thankful you wrote this library and if i can help out with some testing it's the least I can do!
